keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,854 @@
+different user becomes more complicated.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The risks of this proposal are similar to the existing exec credential
+provider.  There are risks associated with execing a binary, for example in a
+scenario where users are downloading binaries from untrustworthy sources to use
+in their cluster.  However, this is not a departure from existing security
+posture of the exec credential feature present in k8s.io/client-go today.
+
+The security and UX will be reviewed by SIG Auth.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Proxy Proposal
+
+The proxy exec plugin is a long-running proxy process which is executed by the
+client on first use.  When executed, the exec proxy plugin will bind and listen
+on a socket, and will a URL on stdout, over which the client would proxy all
+API server bound requests.  The proxy binary location is defined in the
+`ExecConfig` section of a kubeconfig `user`.
+
+The proxy plugin process will be executed from client-go the first time it is
+needed, on transport setup.  If the client fails to connect to the socket
+because no one is listening, client-go will reattempt to exec the plugin.
+Client-go will support both plugins binaries which are themselves the proxy
+process, and additionally plugin binaries which complete, but are responsible
+for starting a proxy process and returning a URL to the proxy.
+
+The client will send its request to the proxy plugin process, and the proxy
+will forward the request to the API server, rather than returning the modified
+request to the client to be sent (see alternatives considered).  The downside
+of this approach is that the proxy must duplicate the some of the transport
+functionality of client packages, like respecting existing proxy configuration,
+but allows for more flexibility for proxy plugin developers.
+
+### Authentication to Proxy
+
+When communicating to the proxy over TCP/IP, the authentication method to the
+proxy is certificate based. The proxy passes back a client certificate and
+certificate authority that the client can use on connection.
+
+When communicating to the proxy over a unix domain socket, the authentication
+method to the proxy is file permission based.  The GID owner of the client
+process is passed to the proxy over stdin, which the proxy should give
+read/write permissions on the socket to.
```

Comment:
I would like to reduce this down to just mTLS over UDS.

Reply:
Done.